sql: simplify and fix error handling from remote flows watcher

yuzefovich opened the pull request:
This commit refactors the way the "remote flows setup watcher" goroutine
communicates an error encountered during `SetupFlowRequest` RPCs
evaluation when they were issued concurrently (i.e. via the DistSQL
workers). This results in code simplification as well as bug fix of
a pretty rare crash that could occur due to a race condition.

As a reminder, during 23.1 cycle in https://github.com/cockroachdb/cockroach/commit/0c1095e31cf93ea7f177f8bf1750ebab188e02d7
we refactored the code for setting up all flows of the distributed plan
so that the gateway flow didn't wait for `SetupFlowRequest` RPCs to come
back from the remote nodes. This required introduction a new "remote
flows setup watcher" goroutine whose only job was to see whether all
RPCs succeed, and if not, to send the error to the main gateway flow
goroutine as well as to cancel that flow. We implemented the
communication of that error by storing it directly into
`rowResultWriter` and introduced mutex protection for synchornization
between two goroutines.

However, that mutex protection was insufficient which could lead to
a crash or a race when reading the error. In particular, the following
could now occur:
1. the main goroutine calls `AddRow` or `AddBatch` concurrently with the
watcher goroutine setting the error. The contract of
`RestrictedCommandResult` is such that most calls after `SetError` are
disallowed, and `AddRow/AddBatch` actually panic in case `err` is set.
This has been observed in a unit test where an error is injected.
2. `AddRow/AddBatch` read `err` field without mutex protection which the
watcher goroutine might be writing into it (under mutex).

These situations are now fixed by making so that only the main goroutine
can set the error on `rowResultWriter`, and we introduce a buffered
channel for the watcher goroutine to send the error on (the error is
picked up on the next call to `Push` or `PushBatch`). This significantly
simplies the code too.

The benchmarks show very minor differences (looks like noise) and are
available [here](https://gist.github.com/yuzefovich/6b4b10d77b2d802ddbcc870eda5f9158).

There is no release note since we haven't seen the bug in the wild, and
it should be quite rare to occur.

Fixes: #111995.

Release note: None

yuzefovich committed:
Revert "sql: switch DistSQLReceiver.status to using atomics"

This reverts commit c371164fcbdcfdb7bbd98c8eedf0095eb0d9341e. It is done
so that the following commit could be easily backported to 23.1 since
c371164fcbdcfdb7bbd98c8eedf0095eb0d9341e is not present there.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/112117)
<!-- Reviewable:end -->

yuzefovich committed:
sql: simplify and fix error handling from remote flows watcher

This commit refactors the way the "remote flows setup watcher" goroutine
communicates an error encountered during `SetupFlowRequest` RPCs
evaluation when they were issued concurrently (i.e. via the DistSQL
workers). This results in code simplification as well as bug fix of
a pretty rare crash that could occur due to a race condition.

As a reminder, during 23.1 cycle in 0c1095e31cf93ea7f177f8bf1750ebab188e02d7
we refactored the code for setting up all flows of the distributed plan
so that the gateway flow didn't wait for `SetupFlowRequest` RPCs to come
back from the remote nodes. This required introduction a new "remote
flows setup watcher" goroutine whose only job was to see whether all
RPCs succeed, and if not, to send the error to the main gateway flow
goroutine as well as to cancel that flow. We implemented the
communication of that error by storing it directly into
`rowResultWriter` and introduced mutex protection for synchornization
between two goroutines.

However, that mutex protection was insufficient which could lead to
a crash or a race when reading the error. In particular, the following
could now occur:
1. the main goroutine calls `AddRow` or `AddBatch` concurrently with the
watcher goroutine setting the error. The contract of
`RestrictedCommandResult` is such that most calls after `SetError` are
disallowed, and `AddRow/AddBatch` actually panic in case `err` is set.
This has been observed in a unit test where an error is injected.
2. `AddRow/AddBatch` read `err` field without mutex protection which the
watcher goroutine might be writing into it (under mutex).

These situations are now fixed by making so that only the main goroutine
can set the error on `rowResultWriter`, and we introduce a buffered
channel for the watcher goroutine to send the error on (the error is
picked up on the next call to `Push` or `PushBatch`). This significantly
simplies the code too.

The benchmarks show very minor differences (looks like noise) and are
available [here](https://gist.github.com/yuzefovich/6b4b10d77b2d802ddbcc870eda5f9158).

There is no release note since we haven't seen the bug in the wild, and
it should be quite rare to occur.

Release note: None

rharding6373 reviewed (commented):
Nice work on the simplification. I have a concern about the potential impact on local reruns.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/112117#-:-NgfAfn59bfP2D38SD3u:b97d5n3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @yuzefovich)

___
*[`pkg/sql/distsql_running.go` line 1277 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/112117#-Ngf8RKv2noV7H-tpnz4:-Ngf8RKv2noV7H-tpnz5:b-t7h0zc) ([raw file](https://github.com/cockroachdb/cockroach/blob/b197aea2ed32819643843f21c82edf7ede39cbba/pkg/sql/distsql_running.go#L1277)):*
> ```Go
> 
> // SetError provides a convenient way for a client to pass in an error, thus
> // pretending that a query execution error happened. The error is passed along
> ```

This comment is confusing for two reasons: 1) This function is only called in distsql_running.go afaict, so who is the client? 2) "pretending" makes it sound like an error didn't happen, which is not the case.
___
*[`pkg/sql/distsql_running.go` line 1329 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/112117#-Ngf6vxX69fv41ipFPFt:-Ngf6vxX69fv41ipFPFu:bw9k3nb) ([raw file](https://github.com/cockroachdb/cockroach/blob/b197aea2ed32819643843f21c82edf7ede39cbba/pkg/sql/distsql_running.go#L1329)):*
> ```Go
> 	}
> 	select {
> 	case err := <-r.concurrentErrorCh:
> ```

Experimenting with channels in the go playground, I noticed that if the channel is closed this assignment will trigger: https://gist.github.com/rharding6373/343f4b1e356008c7592b84630fd5e5f1. I assume that for error the default value would be nil, so it wouldn't overwrite any existing errors?

I'm curious what the impact this has on performance when there's a local rerun.

Do we need to close the channel for local reruns?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/112117)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @rharding6373)

___
*[`pkg/sql/distsql_running.go` line 1277 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/112117#-Ngf8RKv2noV7H-tpnz4:-NguF1V06UcIRST8seS6:bsuohip) ([raw file](https://github.com/cockroachdb/cockroach/blob/b197aea2ed32819643843f21c82edf7ede39cbba/pkg/sql/distsql_running.go#L1277)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

This comment is confusing for two reasons: 1) This function is only called in distsql_running.go afaict, so who is the client? 2) "pretending" makes it sound like an error didn't happen, which is not the case.
</blockquote></details>

This wording was introduced long time ago in 51ca09e2f8023ee1354fa1a14d779f2794a729f3. IIUC the "client" in this case is the user of `DistSQLReceiver` which is the main execution goroutine. "Pretending" here implies that this "client" can inject additional errors that were not the direct result of query computation, and rather were triggered due to some internal execution-infra-specific conditions.

I simplified the comment to remove confusing bits, let me know what you think.
___
*[`pkg/sql/distsql_running.go` line 1329 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/112117#-Ngf6vxX69fv41ipFPFt:-NguKL_44nwtsEPximZZ:b-h695aa) ([raw file](https://github.com/cockroachdb/cockroach/blob/b197aea2ed32819643843f21c82edf7ede39cbba/pkg/sql/distsql_running.go#L1329)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

Experimenting with channels in the go playground, I noticed that if the channel is closed this assignment will trigger: https://gist.github.com/rharding6373/343f4b1e356008c7592b84630fd5e5f1. I assume that for error the default value would be nil, so it wouldn't overwrite any existing errors?

I'm curious what the impact this has on performance when there's a local rerun.

Do we need to close the channel for local reruns?
</blockquote></details>

That's a good catch. I refactored the code to only recv from the channel during distributed runs and skip the channel for local ones. This should have negligible overhead while being simpler to reason about.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice work!

> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/112117#-:-Nh84-1NAQ5a4mDKj8j5:bcmw7ss)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @rharding6373)



<!-- Sent from Reviewable.io -->

rharding6373 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Sorry for the delay.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/112117#-:-Nh84HrlAoYae9QUM2DL:b-8yidtp)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[`pkg/sql/distsql_running.go` line 1277 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/112117#-Ngf8RKv2noV7H-tpnz4:-Nh80iEA6h2vb8eE4-9k:ba23tzz) ([raw file](https://github.com/cockroachdb/cockroach/blob/b197aea2ed32819643843f21c82edf7ede39cbba/pkg/sql/distsql_running.go#L1277)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

This wording was introduced long time ago in 51ca09e2f8023ee1354fa1a14d779f2794a729f3. IIUC the "client" in this case is the user of `DistSQLReceiver` which is the main execution goroutine. "Pretending" here implies that this "client" can inject additional errors that were not the direct result of query computation, and rather were triggered due to some internal execution-infra-specific conditions.

I simplified the comment to remove confusing bits, let me know what you think.
</blockquote></details>

This is less confusing, thanks.
___
*[`pkg/sql/distsql_running.go` line 1329 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/112117#-Ngf6vxX69fv41ipFPFt:-Nh84AIA97fnlmYFL-qg:b-vs1pr4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b197aea2ed32819643843f21c82edf7ede39cbba/pkg/sql/distsql_running.go#L1329)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

That's a good catch. I refactored the code to only recv from the channel during distributed runs and skip the channel for local ones. This should have negligible overhead while being simpler to reason about.
</blockquote></details>

Looks good.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
TFTRs!

bors r+
